python/tskit/combinatorics.py

```diff
@@ -28,12 +28,133 @@
 import functools
 import heapq
 import itertools
+import json
+import random
 
+import attr
 import numpy as np
 
 import tskit
 
 
+@attr.s(eq=False)
+class TreeNode:
+    """
+    Simple linked tree class used to generate tree topologies.
+    """
+
+    parent = attr.ib(default=None)
+    children = attr.ib(factory=list)
+    label = attr.ib(default=None)
+
+    @staticmethod
+    def random_binary_tree(leaf_labels, rng):
+        """
+        Based on the description of Remy's method of generating "decorated"
+        random binary trees in TAOCP 7.2.1.6. This is not a direct
+        implementation of Algorithm R, because we are interested in
+        the leaf node labellings.
+
+        The pre-fasicle text is available here, page 16:
+        http://www.cs.utsa.edu/~wagner/knuth/fasc4a.pdf
+        """
+        nodes = [TreeNode(label=leaf_labels[0])]
+        for label in leaf_labels[1:]:
+            n = TreeNode(label=label)
+            x = rng.choice(nodes)
+            internal = TreeNode(parent=x.parent, children=[x, n])
+            if x.parent is not None:
+                index = x.parent.children.index(x)
+                x.parent.children[index] = internal
+            rng.shuffle(internal.children)
+            x.parent = internal
+            n.parent = internal
+            nodes.extend([n, internal])
+
+        root = nodes[0]
+        while root.parent is not None:
+            root = root.parent
+        return root
+
+
+def split_polytomies(
+    tree,
+    *,
+    epsilon=None,
+    method=None,
+    record_provenance=True,
+    random_seed=None,
+):
+    """
+    Return a new tree where extra nodes and edges have been inserted
+    so that any any node with more than two children is resolved into
+    a binary tree.
+
+    For further documentation, please refer to the :meth:`Tree.split_polytomies`
+    method, which is the usual route through which this function is called.
+    """
+    allowed_methods = ["random"]
+    if epsilon is None:
+        epsilon = 1e-10
+    if method is None:
+        method = "random"
+    if method not in allowed_methods:
+        raise ValueError(f"Method must be chosen from {allowed_methods}")
+
+    tables = tree.tree_sequence.dump_tables()
+    tables.keep_intervals([tree.interval], simplify=False)
```

Comment:
This seems not ideal for a long tree sequence, but I guess there's no easy way around it? Instead of O(E) where E is the total number of edges, seems like there should be an O(K + log(E)) algorithm where K is the number of edges to keep, via stabbing queries? I'm not that familiar with the tskit backend though.

Reply:
This is a great point @brianzhang01. Currently this operation is O(E) as you say, and it's one of the things we could make more efficient if we had implemented stabbing queries. See #684 and other [related issues](https://github.com/tskit-dev/tskit/projects/5).

For this application, I think the current approach is fine, but we would like to speed it up at some point.